File: 6. Delete by position.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, val):
        n = Node(val)
        if not self.head:
            self.head = n
            return
        cur = self.head
        while cur.next:
            cur = cur.next
        cur.next = n

    def delete_pos(self, pos):
        if not self.head:
            return
        if pos <= 0:
            self.head = self.head.next
            return
        cur = self.head
        for _ in range(pos - 1):
            if not cur.next:
                return
            cur = cur.next
        if cur.next:
            cur.next = cur.next.next

    def print_list(self):
        cur = self.head
        out = []
        while cur:
            out.append(str(cur.data))
            cur = cur.next
        print(' -> '.join(out))


if __name__ == '__main__':
    ll = LinkedList()
    ll.append(10)
    ll.append(20)
    ll.append(30)
    ll.append(40)
    print('Before: ', end='')
    ll.print_list()
    ll.delete_pos(2)
    print('After deleting position 2: ', end='')
    ll.print_list()

File: 6. Delete by position.cpp
#include <iostream>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};

void append(Node*& h,int v){ Node* n=new Node(v); if(!h){h=n;return;} Node* c=h; while(c->next) c=c->next; c->next=n; }

void delete_pos(Node*& head,int pos){ if(!head) return; if(pos<=0){ Node* t=head; head=head->next; delete t; return; } Node* cur=head; for(int i=0;i<pos-1 && cur->next;i++) cur=cur->next; if(cur->next){ Node* t=cur->next; cur->next=t->next; delete t; } }
void print(Node* h){ while(h){ cout<<h->data; if(h->next) cout<<" -> "; h=h->next;} cout<<endl; }
int main(){ Node* head=nullptr; append(head,10); append(head,20); append(head,30); append(head,40); cout<<"Before: "; print(head); delete_pos(head,2); cout<<"After deleting position 2: "; print(head); return 0; }
